File: app/src/main/java/com/tak/smsforwarder/DataStore.java
package com.takpack.smsforwarder;

import android.content.Context;
import android.content.SharedPreferences;
import org.json.JSONArray;
import org.json.JSONObject;

public class DataStore {
    private static final String PREF = "tak_sms_store_v7";
    private static final String RULES = "rules";
    private static final String SMS = "sms";
    private static final String ENABLED = "enabled";

    private static SharedPreferences sp(Context c) {
        return c.getSharedPreferences(PREF, Context.MODE_PRIVATE);
    }

    public static void setEnabled(Context c, boolean v) {
        sp(c).edit().putBoolean(ENABLED, v).apply();
    }

    public static boolean isEnabled(Context c) {
        return sp(c).getBoolean(ENABLED, true);
    }

    public static JSONArray getRulesArray(Context c) {
        try { return new JSONArray(sp(c).getString(RULES, "[]")); }
        catch(Exception e) { return new JSONArray(); }
    }

    public static String getRules(Context c) { return getRulesArray(c).toString(); }

    public static boolean saveRule(Context c, String ruleJson) {
        try {
            JSONObject rule = new JSONObject(ruleJson);
            String id = rule.optString("id", "");
            if (id.trim().isEmpty()) {
                id = String.valueOf(System.currentTimeMillis());
                rule.put("id", id);
            }
            if (!rule.has("enabled")) rule.put("enabled", true);
            if (!rule.has("senders")) rule.put("senders", new JSONArray());

            JSONArray oldArr = getRulesArray(c);
            JSONArray newArr = new JSONArray();
            boolean found = false;

            for (int i = 0; i < oldArr.length(); i++) {
                JSONObject old = oldArr.getJSONObject(i);
                if (id.equals(old.optString("id"))) {
                    newArr.put(rule);
                    found = true;
                } else {
                    newArr.put(old);
                }
            }
            if (!found) newArr.put(rule);
            sp(c).edit().putString(RULES, newArr.toString()).apply();
            return true;
        } catch(Exception e) { return false; }
    }

    public static JSONArray getSmsArray(Context c) {
        try { return new JSONArray(sp(c).getString(SMS, "[]")); }
        catch(Exception e) { return new JSONArray(); }
    }

    public static String getSms(Context c) { return getSmsArray(c).toString(); }

    public static void addSms(Context c, String sender, String body, long time) {
        try {
            JSONArray oldArr = getSmsArray(c);
            JSONArray newArr = new JSONArray();

            JSONObject matched = matchRule(c, sender, body);
            JSONObject m = new JSONObject();
            m.put("id", String.valueOf(System.currentTimeMillis()));
            m.put("sender", sender == null ? "" : sender);
            m.put("body", body == null ? "" : body);
            m.put("time", time);
            m.put("rule", matched.optString("rule", "عمومی"));
            m.put("senderNote", matched.optString("senderNote", ""));
            newArr.put(m);

            int limit = Math.min(oldArr.length(), 99);
            for (int i = 0; i < limit; i++) newArr.put(oldArr.getJSONObject(i));
            sp(c).edit().putString(SMS, newArr.toString()).apply();
        } catch(Exception ignored) {}
    }

    private static String normalize(String x) {
        if (x == null) return "";
        return x.replace(" ", "").replace("-", "").replace("+98", "0").trim().toLowerCase();
    }

    public static JSONObject matchRule(Context c, String sender, String body) {
        JSONObject result = new JSONObject();
        try {
            String s = sender == null ? "" : sender;
            String b = body == null ? "" : body;
            String ns = normalize(s);
            String all = (s + " " + b).toLowerCase();

            JSONArray rules = getRulesArray(c);
            for (int i = 0; i < rules.length(); i++) {
                JSONObject r = rules.getJSONObject(i);
                if (!r.optBoolean("enabled", true)) continue;

                boolean senderOk = false;
                String senderNote = "";
                JSONArray senders = r.optJSONArray("senders");

                if (senders == null || senders.length() == 0) {
                    senderOk = true;
                } else {
                    for (int j = 0; j < senders.length(); j++) {
                        JSONObject it = senders.getJSONObject(j);
                        String n = normalize(it.optString("number", ""));
                        if (n.length() == 0 || ns.contains(n) || n.contains(ns)) {
                            senderOk = true;
                            senderNote = it.optString("note", "");
                            break;
                        }
                    }
                }

                String keywords = r.optString("keywords", "").trim().toLowerCase();
                boolean keyOk = keywords.isEmpty() || "همه".equals(keywords);
                if (!keyOk) {
                    String[] parts = keywords.split("[,،\\s]+");
                    for (String p : parts) {
                        p = p.trim();
                        if (p.length() > 0 && all.contains(p)) { keyOk = true; break; }
                    }
                }

                if (senderOk && keyOk) {
                    result.put("rule", r.optString("name", "قانون ذخیره‌شده"));
                    result.put("senderNote", senderNote);
                    return result;
                }
            }

            if (all.contains("بانک") || all.contains("کارت") || all.contains("تراکنش")) result.put("rule", "بانک‌ها");
            else if (all.contains("otp") || all.contains("کد") || all.contains("رمز")) result.put("rule", "OTP");
            else if (all.contains("تخفیف") || all.contains("فروش")) result.put("rule", "تبلیغات");
            else result.put("rule", "عمومی");
            result.put("senderNote", "");
        } catch(Exception ignored) {}
        return result;
    }
}

File: app/src/main/java/com/tak/smsforwarder/DataStore.kt
package com.takpack.smsforwarder

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

object DataStore {
    private const val PREF = "tak_sms_forwarder_kotlin_final_fixed_v31"
    private const val RULES = "rules"
    private const val MESSAGES = "messages"
    private const val SETTINGS = "settings"
    private const val ENABLED = "enabled"

    private fun sp(context: Context) = context.getSharedPreferences(PREF, Context.MODE_PRIVATE)

    fun setEnabled(context: Context, enabled: Boolean) {
        sp(context).edit().putBoolean(ENABLED, enabled).apply()
    }

    fun isEnabled(context: Context): Boolean = sp(context).getBoolean(ENABLED, true)

    fun getRulesArray(context: Context): JSONArray {
        return try { JSONArray(sp(context).getString(RULES, "[]")) } catch (_: Exception) { JSONArray() }
    }

    fun getRules(context: Context): String = getRulesArray(context).toString()

    private fun inferForwardType(rule: JSONObject): String {
        val hasSms = rule.optString("smsTarget", "").trim().isNotBlank()
        val hasEmail = rule.optString("emailTarget", "").trim().isNotBlank()
        return when {
            hasSms && hasEmail -> "both"
            hasSms -> "sms"
            hasEmail -> "email"
            else -> "history"
        }
    }

    fun saveRule(context: Context, rawRule: String): Boolean {
        return try {
            val rule = JSONObject(rawRule)
            var id = rule.optString("id", "")
            if (id.isBlank()) {
                id = System.currentTimeMillis().toString()
                rule.put("id", id)
            }
            if (!rule.has("enabled")) rule.put("enabled", true)
            if (!rule.has("senders")) rule.put("senders", JSONArray())
            rule.put("forwardType", inferForwardType(rule))

            val old = getRulesArray(context)
            val next = JSONArray()
            var found = false

            for (i in 0 until old.length()) {
                val item = old.getJSONObject(i)
                if (item.optString("id") == id) {
                    next.put(rule)
                    found = true
                } else {
                    next.put(item)
                }
            }

            if (!found) next.put(rule)
            sp(context).edit().putString(RULES, next.toString()).apply()
            true
        } catch (_: Exception) {
            false
        }
    }

    fun deleteRule(context: Context, id: String): Boolean {
        return try {
            val old = getRulesArray(context)
            val next = JSONArray()
            for (i in 0 until old.length()) {
                val item = old.getJSONObject(i)
                if (item.optString("id") != id) next.put(item)
            }
            sp(context).edit().putString(RULES, next.toString()).apply()
            true
        } catch (_: Exception) {
            false
        }
    }

    fun toggleRule(context: Context, id: String, enabled: Boolean): Boolean {
        return try {
            val arr = getRulesArray(context)
            for (i in 0 until arr.length()) {
                val item = arr.getJSONObject(i)
                if (item.optString("id") == id) item.put("enabled", enabled)
            }
            sp(context).edit().putString(RULES, arr.toString()).apply()
            true
        } catch (_: Exception) {
            false
        }
    }

    fun getSettingsObject(context: Context): JSONObject {
        return try {
            JSONObject(sp(context).getString(SETTINGS, "{}") ?: "{}")
        } catch (_: Exception) {
            JSONObject()
        }
    }

    fun getSettings(context: Context): String = getSettingsObject(context).toString()

    fun saveSettings(context: Context, raw: String): Boolean {
        return try {
            val obj = JSONObject(raw)
            sp(context).edit().putString(SETTINGS, obj.toString()).apply()
            true
        } catch (_: Exception) {
            false
        }
    }

    fun getMessagesArray(context: Context): JSONArray {
        return try { JSONArray(sp(context).getString(MESSAGES, "[]")) } catch (_: Exception) { JSONArray() }
    }

    fun getMessages(context: Context): String = getMessagesArray(context).toString()

    fun clearMessages(context: Context) {
        sp(context).edit().putString(MESSAGES, "[]").apply()
    }

    fun deleteMessage(context: Context, id: String): Boolean {
        return try {
            val old = getMessagesArray(context)
            val next = JSONArray()
            for (i in 0 until old.length()) {
                val item = old.getJSONObject(i)
                if (item.optString("id") != id) next.put(item)
            }
            sp(context).edit().putString(MESSAGES, next.toString()).apply()
            true
        } catch (_: Exception) { false }
    }

    fun addMessage(context: Context, sender: String, body: String, time: Long, rule: JSONObject, status: String, error: String) {
        try {
            val old = getMessagesArray(context)
            val next = JSONArray()
            val msg = JSONObject()
            msg.put("id", System.currentTimeMillis().toString() + "_" + kotlin.random.Random.nextInt(1000, 9999))
            msg.put("sender", sender)
            msg.put("body", body)
            msg.put("time", time)
            msg.put("ruleId", rule.optString("id", ""))
            msg.put("ruleName", rule.optString("name", "بدون نام"))
            msg.put("senderNote", RuleMatcher.senderNote(rule, sender))
            msg.put("forwardType", rule.optString("forwardType", "history"))
            msg.put("forwardTarget", RuleMatcher.forwardTarget(rule))
            msg.put("status", status)
            msg.put("error", error)
            next.put(msg)
            val limit = minOf(old.length(), 199)
            for (i in 0 until limit) next.put(old.getJSONObject(i))
            sp(context).edit().putString(MESSAGES, next.toString()).apply()
        } catch (_: Exception) {}
    }

    fun exportAll(context: Context): String {
        return JSONObject()
            .put("app", "TAK SMS Forwarder")
            .put("version", 1)
            .put("createdAt", System.currentTimeMillis())
            .put("rules", getRulesArray(context))
            .put("messages", getMessagesArray(context))
            .put("settings", getSettingsObject(context))
            .put("enabled", isEnabled(context))
            .toString(2)
    }

    fun importAll(context: Context, raw: String): Boolean {
        return try {
            val obj = JSONObject(raw)
            val rules = obj.optJSONArray("rules") ?: JSONArray()
            val messages = obj.optJSONArray("messages") ?: JSONArray()
            val settings = obj.optJSONObject("settings") ?: JSONObject()
            val enabled = obj.optBoolean("enabled", true)
            sp(context).edit()
                .putString(RULES, rules.toString())
                .putString(MESSAGES, messages.toString())
                .putString(SETTINGS, settings.toString())
                .putBoolean(ENABLED, enabled)
                .apply()
            true
        } catch (_: Exception) { false }
    }
}
